### PR TITLE
[CI] Run ITs for the latest releases and snapshots

### PR DESCRIPTION
**Pull request by v1v**
### What

Run ITs for the Elastic versions:

- 7.13.1 release
- 7.13-SNAPSHOT

`8.0.0-SNAPSHOT` has been disabled since the APM Server and Fleet Server are not deployed.

Minor changes in the wait for the enrolment to be available and provide GH checks with the status of each stage in the CI.

### Actions

- Automate the bump of the elastic versions when a new release is published.

### Issues

Depends on https://github.com/elastic/azure-vm-extension/pull/12

### Tests

Automate the version

```
➜  azure-vm-extension git:(feature/matrix) ✗ cat .ci/Jenkinsfile | grep 8.0
            values '8.0.0-SNAPSHOT', '7.13.1-SNAPSHOT', '7.13.1'
➜  azure-vm-extension git:(feature/matrix) ✗ sed -E -e "s#(values '8.0.0-SNAPSHOT',) '[0-9]+\.[0-9]+\.[0-9]+-SNAPSHOT', '[0-9]+\.[0-9]+\.[0-9]+'#\1: '${VERSION_RELEASE}-SNAPSHOT', '${VERSION_DEV}'#g" .ci/Jenkinsfile | grep 8.0
            values '8.0.0-SNAPSHOT',: '1.2.3-SNAPSHOT', '3.2.1'
```

### CODE REVIEW

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -135,82 +151,106 @@ def destroyTerraform( ) {
...
+    def username = value?.username
+    def password = value?.password
+    if(cloud_id == null || username == null || password == null){
+      error "withCloudEnv: was not possible to get the authentication info."
+    }
+    withEnvMask(vars: [
+      [var: 'TF_VAR_username', password: username],
+      [var: 'TF_VAR_password', password: password],
+      [var: 'TF_VAR_cloudId', password: cloud_id],
+      [var: 'TF_VAR_password', password: password]
```
Comment (v1v) [inline .ci/Jenkinsfile:211]:
```suggestion
      [var: 'TF_VAR_cloudId', password: cloud_id]
```

File: .ci/Jenkinsfile [hunk 2]
```diff
@@ -135,82 +162,105 @@ def destroyTerraform( ) {
...
-    [var: 'AZ_TENANT', password: tenant],
-    [var: 'AZ_SUBSCRIPTION', password: subscription]
-  ]){
+}
+
+def withMatrixEnv(Closure body) {
+  // -S need to avoid clashes between same version for releases and snapshots.
+  // 7.13.0-SNAPSHOT and 7.13.0 might clash so let's keep the -S for this.
+  def stackVersion = env.ELASTIC_STACK_VERSION.replaceAll('-SNAPSHOT','-S').replaceAll('\\.', '-')
+  def uniqueId = "${stackVersion}-${BUILD_ID}-${BRANCH_NAME}"
```
Comment (v1v) [inline .ci/Jenkinsfile:257]: This will be refactored with https://github.com/elastic/apm-pipeline-library/pull/1147

File: test/ats/validate.sh [hunk 3]
```diff
@@ -39,6 +39,8 @@ function search() {
     --connect-timeout 30 \
     --retry 30 \
     --retry-delay 10 \
+    --retry-all-errors \
+    --fail \
```
Comment (v1v) [inline test/ats/validate.sh:43]: This is a leftover from the previous execution to support 5xx and 4xx http error codes

File: .ci/Jenkinsfile [hunk 4]
```diff
@@ -135,82 +164,106 @@ def destroyTerraform( ) {
...
+  withEnv([
+    "CLUSTER_NAME=tst-az-${BUILD_ID}-${BRANCH_NAME}-${ELASTIC_STACK_VERSION}",
+    'TF_VAR_prefix=' + uniqueId.take(9) + '0',
+    'TF_VAR_vmName=' + uniqueId.take(14) + '0'
```
Comment (v1v) [inline .ci/Jenkinsfile:263]:
```suggestion
    'TF_VAR_prefix=' + randomString(size: 10),
    'TF_VAR_vmName=' + randomString(size: 15)
```

File: .ci/Jenkinsfile [hunk 5]
```diff
@@ -135,82 +164,106 @@ def destroyTerraform( ) {
...
-    [var: 'AZ_TENANT', password: tenant],
-    [var: 'AZ_SUBSCRIPTION', password: subscription]
-  ]){
+}
+
+def withMatrixEnv(Closure body) {
+  // -S need to avoid clashes between same version for releases and snapshots.
+  // 7.13.0-SNAPSHOT and 7.13.0 might clash so let's keep the -S for this.
+  def stackVersion = env.ELASTIC_STACK_VERSION.replaceAll('-SNAPSHOT','-S').replaceAll('\\.', '-')
+  def uniqueId = "${stackVersion}-${BUILD_ID}-${BRANCH_NAME}"
```
Comment (v1v) [inline .ci/Jenkinsfile:259]:
```suggestion
```

File: .ci/Jenkinsfile [hunk 6]
```diff
@@ -41,76 +35,109 @@ pipeline {
...
+      failFast false
+      matrix {
+        agent { label 'ubuntu-20' }
+        axes {
+          axis {
+            name 'ELASTIC_STACK_VERSION'
+            // The below line is part of the bump release automation
+            // if you change anything please modifies the file
+            // .ci/bump-stack-release-version.sh
+            values '8.0.0-SNAPSHOT', '7.14.0-SNAPSHOT', '7.13.1'
```
Comment (v1v) [inline .ci/Jenkinsfile:53]:
```suggestion
            // 7.x branch is a special one that requires to use https://github.com/elastic/apm-pipeline-library/pull/1149
            values '8.0.0-SNAPSHOT', '7.x', '7.13.1'
```

File: .ci/Jenkinsfile [hunk 7]
```diff
@@ -135,82 +170,106 @@ def destroyTerraform( ) {
...
+    if (props?.errors) {
+      error "withAzEnv: Unable to get credentials from the vault: ${props.errors.toString()}"
+    }
+    def value = props?.data
+    def tenant = value?.tenant
+    def username = value?.username
+    def password = value?.password
+    def subscription = value?.subscription
+    if(tenant == null || username == null || password == null || subscription == null){
+      error "withAzEnv: was not possible to get the authentication info."
```
[the remark this answers is not in the document]
Comment (v1v) [reply .ci/Jenkinsfile:248]: I'll move all this withXXX methods to the shared library, for two reason:
- simplify the pipeline
- add UTs and the validation to be more specific.

And example -> https://github.com/elastic/apm-pipeline-library/pull/1150

File: .ci/Jenkinsfile (hunk 4 above)
Comment (v1v) [reply .ci/Jenkinsfile:269]: Requires https://github.com/elastic/apm-pipeline-library/pull/1147

**Review by cachedout — APPROVED** [review]

**Comment by v1v** [comment]
I'll refactor in a follow up all the steps that were moved to the shared library

**Pull request merged**